[TCSACR-399] Add vision pose detection doc (#1567)
Samsung/tizen-docs: docs/application/dotnet/guides/multimedia/pose-detection.md

# Pose Detection

> Pose Detection is a new feature of Media Vision Inference API since Tizen 6.5(C#). This feature provides landmark detection. Besides, it defines landmarks and parts of a human body to help detect a human pose with Motion Capture (MoCap) file, which you can create or edit using various tools.

Reviewer: Is the "(C#)" relevant here?
```suggestion
Pose Detection is a new feature of Media Vision Inference API since Tizen 6.5(C#). This feature provides landmark detection; it defines landmarks and parts of a human body to help detect a human pose with a Motion Capture (MoCap) file, which you can create or edit using various tools.
```

Author: Yes. C API support since Tizen 6.0

**Figure: Definition of human body pose landmarks and body parts**

![Body pose](/application/native/guides/multimedia/media/mediavision_pose_tizen_def.png)

The Pose landmark detection models are available in Open Model Zoo such as [hosted model zoo](https://www.tensorflow.org/lite/guide/hosted_models#floating_point_models) or public GitHub site such as [public pose model](https://github.com/tyoungroy/PoseEstimationForMobile). The public pose models provide landmark information, such as the number of landmarks and locations. To use them correctly, you must map the information to landmarks based on the definition. For example, you can use the [public pose model](https://github.com/tyoungroy/PoseEstimationForMobile), which provides 14 landmarks as follows:

![Body pose](/application/native/guides/multimedia/media/mediavision_pose_public_model_def.png),

In this model, `-1` denotes that there are no landmarks. Using this landmark information, you can create a mapping file. Suppose you create a mapping file with the name `pose_mapping.txt`, then you can populate the `pose_mapping.txt` file as follows:

```
1
2
-1
3
4
5
6
7
8
-1
9
10
11
12
13
14
```

> `1` denotes that the first landmark of the model corresponds to the first definition,  `MV_INFERENCE_HUMAN_POSE_HEAD`. `-1` at third position denotes that there is no corresponding landmark `MV_INFERENCE_HUMAN_POSE_THORAX`. `3` at the fourth position denotes that the third landmark of the model corresponds to the fourth, `MV_INFERENCE_HUMAN_POSE_RIGHT_SHOULDER`. The following table shows how the public model works:

Reviewer: ```suggestion
`1` denotes that the first landmark of the model corresponds to the first definition,  `MV_INFERENCE_HUMAN_POSE_HEAD`. `-1` at the third position denotes that there is no corresponding landmark `MV_INFERENCE_HUMAN_POSE_THORAX`. `3` at the fourth position denotes that the third landmark of the model corresponds to the fourth, `MV_INFERENCE_HUMAN_POSE_RIGHT_SHOULDER`. The following table shows how the public model works:
```

![Body pose](/application/native/guides/multimedia/media/mediavision_pose_bvh_sample.png)

The example starts with hips and ends with the left foot with 15 landmarks. You can create a mapping file named `mocap_mapping.txt` as follows:

```
Hips,10
Neck,2
Head,1
LeftUpArm,7
LeftLowArm,8
LeftHand,9
RightUpArm,4
RightLowArm,5
RightHand,6
LeftUpLeg,14
LeftLowLeg,15
LeftFoot,16
RightUpLeg,11
RightLowLeg,12
RightFoot,13
```
If there is no mapped landmark, you don't need to list it. For example, index three is missing. It means that, in this BVH file, the third definition, `MV_INFERENCE_HUMAN_POSE_THORAX` is not defined and not used.

## Prerequisites

> To enable your application to use the media vision inference functionality:

Reviewer: ```suggestion
To enable your application to use the Media Vision Inference functionality:
```

1.  Install the NuGet packages for Media Vision.

    ```csharp
    using Tizen.Multimedia;
    using Tizen.Multimedia.Vision;
    ```
2. Create a structure to store pose data.

   For pose detection, use the following `BodyPart` structure:

   ```csharp
   public enum BodyPart
   {
       Head,
       Neck,
       RightUpArm,
       RightLowArm,
       RightHand,
       LeftUpArm,
       LeftLowArm,
       LeftHand,
       RightUpLeg,
       RightLowLeg,
       RightFoot,
       LeftUpLeg,
       LeftLowLeg,
       LeftFoot,
   }
   ```

## Detect human pose

To detect human pose from an image:

1. Create the source and engine configuration handles:

   ```csharp
   pdConfig = new InferenceModelConfiguration();
   pdConfig.WeightFilePath = Application.Current.DirectoryInfo.Resource + "model.tflite";
   pdConfig.MetadataFilePath = Application.Current.DirectoryInfo.Resource + "model.json";
   pdConfig.Backend = InferenceBackendType.TFLite;
   pdConfig.Device = InferenceTargetDevice.CPU;
   pdConfig.LoadInferenceModel();
   ```

> 2. Get image file and fill the MediaVisionSource `source` with the decoded raw data.

Reviewer: ```suggestion
2. Get the image file and fill the MediaVisionSource `source` with the decoded raw data.
```

   In the following example image `sample.jpg`, a person is shown in a squat pose:

   <img alt="sample.jpg" src="/application/native/guides/multimedia/media/mediavision_pose_sample_sumo.png" width=300>

   ```csharp
   MediaVisionSource source = new MediaVisionSource(rgbframe, width, height, Tizen.Multimedia.ColorSpace.Rgb888);
   ```

> 3. To detect landmark of the pose from the `sample.jpg` image, run a `DetectAsync` from Detector.

Reviewer: ```suggestion
3. To detect landmarks of the pose from the `sample.jpg` image, run a `DetectAsync` from Detector.
```

   This will return `Landmark`:

   ```csharp
   Task<Landmark[,]> taskLandmark = Tizen.Multimedia.Vision.PoseLandmarkDetector.DetectAsync(source, pdConfig);
   ```

4. Use `Landmark` to get landmarks information from the image:

>    Landmark is two dimension array which represents number of source for first dimension and enum `BodyPart` for second dimension.

Reviewer: ```suggestion
   Landmark is a two-dimension array that represents the number of sources for the first dimension and enum `BodyPart` for the second dimension.
```

    ```csharp
    public struct Landmark
    {
        /// <summary>
        /// Represents a location in the 2D space.
        /// </summary>
        /// <since_tizen> 9 </since_tizen>
        public Point Location
        {
            get;
            set;
        }
        /// <summary>
        /// Confidence score of point.
        /// </summary>
        /// <since_tizen> 9 </since_tizen>
        public float Score
        {
            get;
            set;
        }
    }
    ```

## Related information

- Dependencies

>   - Tizen 6.5 and Higher for Mobile

Reviewer: ```suggestion
  - Tizen 6.5 or higher for Mobile
```

>   - Tizen 6.5 or higher

Reviewer 2: @KwangSon, @asm-foysal, I would recommend to do the following change for consistency sake with other Dotnet guides.

```suggestion
  - Tizen 6.5 and Higher
```

Reviewer 3: Thanks for checking. Also please don't forget to revert the change done to dotnet/guides/multimedia/media-vision.md above.

>   - Tizen 6.5 or higher for Mobile

Reviewer 2: For .Net guides we don't mention Mobile/Wearable, so kindly change to
```suggestion
  - Tizen 6.5 or higher
```

Reviewer 3: Question. This `Tizen X.X and Higher` is used to other all pages, only in dotnet about 150 pages .

Are you going to change the style to use small letter to "higher",
and update every pages including dotnet and native and web?
If not, how about following the existing style, as guided to you with https://github.com/Samsung/tizen-docs/blob/52cdb00867da738ebd15afc4c5cce88530cb993b/reviewguide/review_points_guide.md#related_info_dependencies?

Ref history : https://github.com/Samsung/tizen-docs/pull/1567#discussion_r764821794

```
@asm-foysal asm-foysal 5 days ago
Suggested change 
  - Tizen 6.5 and Higher for Mobile
  - Tizen 6.5 or higher for Mobile
```

Author: @mijin85cho  capital 'Higher' makes sense. I'll commit.
but another issue is 'and' 'or' case. I guess 'Tizen 6.5 or Higher' is correct. please confirm.

Reviewer 3: As Tizen 6.5 is included in supported version, I think 'and' is proper.
Doesn't 'or' have meaning of "only Tizen 6.5 is supported, otherwise only higher version(ex Tizen 7.0....etc except Tizen 6.5) is supported"?
